src/mini_s3_signing.erl

```diff
@@ -0,0 +1,529 @@
+%% -*- erlang-indent-level: 4;indent-tabs-mode: nil; fill-column: 92 -*-
+%% ex: ts=4 sw=4 et
+%% Amazon Simple Storage Service (S3)
+%% Copyright 2010-2019 Chef, Inc. All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+
+-module(mini_s3_signing).
+
+-export([s3_url/6,
+         s3_request/8,
+         make_authorization/10,
+         universaltime/0,
+         make_v4_canonical_request/7]).
+
+-export([canonicalize_headers/1,
+         retrieve_header_value/2,
+         make_signed_url_authorization_v2/5,
+         make_signed_url_authorization_v4/5
+        ]).
+
+-export_type([config/0]).
+
+
+-ifdef(TEST).
+-compile([export_all]).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-include("internal.hrl").
+-include_lib("xmerl/include/xmerl.hrl").
+
+-opaque config() :: #config{}.
+
+-define(V4_SIGNING_ALGORITHM_NAME, "AWS4-HMAC-SHA256").
+-define(S3_SCOPE_NAME, "s3"). % is there a better name for this?
+-define(V4_SCOPE_TERMINATION, "aws4_request").
+-define(UNSIGNED_PAYLOAD, "UNSIGNED-PAYLOAD").
+
+-define(XMLNS_S3, "http://s3.amazonaws.com/doc/2006-03-01/").
+
+
+%% @doc Canonicalizes a proplist of {"Header", "Value"} pairs by
+%% lower-casing all the Headers.
+%%
+-spec canonicalize_headers([{string() | binary() | atom(), Value::string()}]) ->
+                                  [{LowerCaseHeader::string(), Value::string()}].
+canonicalize_headers(Headers) ->
+    [ {string:to_lower(to_string(H)), V} || {H, V} <- Headers ].
+
+
+%% AWS V4 requires sorting;
+%% Each header is terminated with a "\n"
+make_header_string(Headers) ->
```

Comment:
Nit: Arguably this could called from `canonicalize_headers`' instead of `make_v4_canonical_request` since as amazon describes the canonical header it itself as  both sorted and lowercase